### PR TITLE
Bug 195

### PR DESCRIPTION
**Pull request by zalmen**

### CODE REVIEW

File: p2p/connectionpool/connectionpool.go [hunk 1]
```diff
@@ -96,27 +97,56 @@ func (cp *ConnectionPool) handleDialResult(rPub crypto.PublicKey, result dialRes
...
 	}
-	cp.connections[rPub.String()] = conn
+	cp.connections[rPub.String()] = newConn
 	cp.connMutex.Unlock()
 
 	// update all registered channels
-	res := dialResult{conn, nil}
+	res := dialResult{newConn, nil}
 	cp.handleDialResult(rPub, res)
+	return
```
Comment (y0sher) [inline p2p/connectionpool/connectionpool.go]: no need to return here.

**Review by y0sher — COMMENTED** [review]
looks good to me.

**Review by y0sher — APPROVED** [review]

File: p2p/connectionpool/connectionpool.go [hunk 2]
```diff
@@ -96,27 +97,56 @@ func (cp *ConnectionPool) handleDialResult(rPub crypto.PublicKey, result dialRes
...
+		if res <= 0 { // newConn >= curConn
+			if res == 0 { // newConn == curConn
+				// TODO Is it a potential threat (session hijacking)? Should we keep the existing connection?
+				cp.net.Logger().Warning("new connection was created with same session ID as an existing connection, keeping the new connection (assuming existing connection is stale). existing session ID=%v, new session ID=%v, remote=%s", curConn.Session().ID(), newConn.Session().ID(), rPub)
+			} else {
+				cp.net.Logger().Info("connection created while connection already exists between peers, closing existing connection. existing session ID=%v, new session ID=%v, remote=%s", curConn.Session().ID(), newConn.Session().ID(), rPub)
+			}
+			closeConn = curConn
+			cp.connections[rPub.String()] = newConn
+			// we don't need to update on the new connection since there were already a connection in the table and there shouldn't be any registered channel waiting for updates
```
Comment (y0sher) [inline p2p/connectionpool/connectionpool.go]: this wasn't clear at first, maybe move it to where we return

File: p2p/connectionpool/connectionpool.go [hunk 3]
```diff
@@ -96,26 +97,55 @@ func (cp *ConnectionPool) handleDialResult(rPub crypto.PublicKey, result dialRes
...
+func compareConnections(conn1 net.Connection, conn2 net.Connection) int {
+
+	return bytes.Compare(conn1.Session().ID(), conn2.Session().ID())
+}
+
+func (cp *ConnectionPool) handleNewConnection(rPub crypto.PublicKey, newConn net.Connection, source net.ConnectionSource) {
 	cp.connMutex.Lock()
-	cp.net.Logger().Info("new connection %v -> %v. id=%s", cp.localPub, rPub, conn.ID())
+	var srcPub, dstPub string
+	if source == net.Local {
```
Comment (gavraz) [inline p2p/connectionpool/connectionpool.go:108]: IMO, in simple assignments a multi-assignment is cleaner
src, dst = X, Y

File: p2p/connectionpool/connectionpool.go [hunk 4]
```diff
@@ -96,26 +97,55 @@ func (cp *ConnectionPool) handleDialResult(rPub crypto.PublicKey, result dialRes
...
 	if ok {
+		// it is possible to get a new connection with the same peers as another existing connection, in case the two peers tried to connect to each other at the same time.
+		// We need both peers to agree on which connection to keep and which one to close otherwise they might end up closing both connections (bug #195)
+		res := compareConnections(curConn, newConn)
+		var closeConn net.Connection
+		if res <= 0 { // newConn >= curConn
+			if res == 0 { // newConn == curConn
+				// TODO Is it a potential threat (session hijacking)? Should we keep the existing connection?
+				cp.net.Logger().Warning("new connection was created with same session ID as an existing connection, keeping the new connection (assuming existing connection is stale). existing session ID=%v, new session ID=%v, remote=%s", curConn.Session().ID(), newConn.Session().ID(), rPub)
+			} else {
```
Comment (gavraz) [inline p2p/connectionpool/connectionpool.go:127]: The else is unnecessary.

File: p2p/connectionpool/connectionpool_test.go [hunk 5]
```diff
@@ -61,68 +61,98 @@ func TestGetConnectionWithError(t *testing.T) {
...
+	waitCh := make(chan net.Connection)
+	// dispatch 2 GetConnection calls
+	dispatchF := func(ch chan net.Connection) {
+		conn, _ := cPool.GetConnection(addr, remotePub)
+		assert.Equal(t, remotePub.String(), conn.RemotePublicKey().String())
+		ch <- conn
+	}
+	go dispatchF(waitCh)
+	go dispatchF(waitCh)
+	cnt := 0
```
Comment (gavraz) [inline p2p/connectionpool/connectionpool_test.go:81]: What are we counting? maybe rename it...

File: p2p/connectionpool/connectionpool_test.go [hunk 6]
```diff
@@ -61,68 +61,98 @@ func TestGetConnectionWithError(t *testing.T) {
...
+	var prevId string
+Loop:
+	for {
+		select {
+		case c := <-waitCh:
```
Comment (gavraz) [inline p2p/connectionpool/connectionpool_test.go:86]: Rename to conn?

**Pull request merged**